### PR TITLE
feat(payment-processor): batch conversion - erc20

### PR DESCRIPTION
**Pull request by olivier7delf**
## Implement payment processor for batch conversion proxy

It implements only the batchRouter function, which allows paying within a single transaction N heterogeneous (DAI-DAI, or EUR-DAI) payments.
Todo: implement the other method later, there have not been prioritized yet.

### CODE REVIEW

File: packages/payment-processor/test/payment/shared.ts [hunk 1]
```diff
@@ -24,3 +24,10 @@ export const currencyManager = new CurrencyManager([
     type: RequestLogicTypes.CURRENCY.ERC20,
   })),
 ]);
+
+export const sameCurrencyValue = (
```
Comment (leoslr) [inline packages/payment-processor/test/payment/shared.ts:28]: Is this used somewhere?

**Review by leoslr — APPROVED** [review]

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+ */
+export function encodePayBatchConversionRequest(enrichedRequests: EnrichedRequest[]): string {
+  const { feeAddress } = getRequestPaymentValues(enrichedRequests[0].request);
+
+  let firstPn0Extension: IState<any> | undefined;
+  const pn2requests: ClientTypes.IRequestData[] = [];
+  const conversionDetails: PaymentTypes.ConversionDetail[] = [];
+
+  // fill conversionDetails and pn2requests
+  for (let i = 0; i < enrichedRequests.length; i++) {
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:81]: indeed!

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+      BATCH_PAYMENT_NETWORK_ID.BATCH_MULTI_ERC20_CONVERSION_PAYMENTS
+    ) {
+      firstPn0Extension =
+        firstPn0Extension ?? getPaymentNetworkExtension(enrichedRequests[i].request);
+
+      comparePnTypeAndVersion(firstPn0Extension, enrichedRequests[i].request);
+      if (
+        // the type used must a fiat or an ERC20
+        ![RequestLogicTypes.CURRENCY.ISO4217, RequestLogicTypes.CURRENCY.ERC20].includes(
+          enrichedRequests[i].request.currencyInfo.type,
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:93]: I made a proposition, but I am not sure I get your point.
`const isErc20Currency =
        enrichedRequest.request.currencyInfo.type === RequestLogicTypes.CURRENCY.ERC20;
      const isISO4217Currency =
        enrichedRequest.request.currencyInfo.type === RequestLogicTypes.CURRENCY.ISO4217;
      // the type used must a fiat or an ERC20
      if (!(isErc20Currency || isISO4217Currency))
        throw new Error(`wrong request currencyInfo type`);`

File: packages/payment-processor/src/payment/index.ts [hunk 4]
```diff
@@ -326,3 +326,18 @@ const throwIfNotWeb3 = (request: ClientTypes.IRequestData) => {
     throw new UnsupportedPaymentChain(request.currencyInfo.network);
   }
 };
+
+/**
+ * Input of batch conversion payment processor
+ * It contains requests, paymentSettings, amount and feeAmount.
+ * Currently, these requests must have the same PN, version, and batchFee
+ * Also used in Invoicing repository.
+ * @dev next step: paymentNetworkId could get more values options, see the "ref"
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/index.ts:335]: Yes, you can conversion and no conversion in a single transaction.

Implementing "paymentNetworkId = 1" increase a bit the complexity for very few benefits in gas.
I think we should implement this, and also the ETH-ETH and Fiat-ETH payment network later (the smart contract allows it)

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+          'request cannot be processed, or is not an pn-any-to-erc20-proxy request',
+        );
+      });
+      it("should throw an error if one request's currencyInfo has no value", async () => {
+        EURRequest.currencyInfo.value = '';
+        await expect(
+          payBatchConversionProxyRequest(enrichedRequests, batchConvVersion, wallet),
+        ).rejects.toThrowError("The currency '' is unknown or not supported");
+      });
+      it('should throw an error if request has no extension', async () => {
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:296]: right!

File: packages/payment-processor/test/payment/shared.ts (hunk 1 above)
Comment (olivier7delf) [reply packages/payment-processor/test/payment/shared.ts:28]: nope, cleaned

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts]: No, I meant thera are functions called isErc20Currency & isISO4217Currency that can be used like this:
```ts
if(!(isErc20Currency(request.currencyInfo) || isISO4217Currency(request.currencyInfo)))
```

File: packages/payment-processor/src/payment/batch-proxy.ts [hunk 6]
```diff
@@ -116,9 +116,7 @@ export function encodePayBatchRequest(requests: ClientTypes.IRequestData[]): str
     const pn = getPaymentNetworkExtension(requests[0]);
     for (let i = 0; i < requests.length; i++) {
       validateErc20FeeProxyRequest(requests[i]);
-      if (!comparePnTypeAndVersion(pn, requests[i])) {
-        throw new Error(`Every payment network type and version must be identical`);
-      }
+      comparePnTypeAndVersion(pn, requests[i]);
```
Comment (bertux) [inline packages/payment-processor/src/payment/batch-proxy.ts:119]: are you sure changing how this function manages errors is not breaking something in other projects ?

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+
+// requests setting
+
+const EURExpectedAmount = 100;
+const EURFeeAmount = 2;
+// amounts used for DAI and FAU requests
+const expectedAmount = 100000;
+const feeAmount = 100;
+
+const validEURRequest: ClientTypes.IRequestData = {
```
Comment (bertux) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:61]: name EURValidRequest to make it coherent with other names

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+  beforeAll(async () => {
+    // Revoke DAI and FAU approvals
+    await revokeErc20Approval(
+      getBatchConversionProxyAddress(validEURRequest, batchConvVersion),
+      DAITokenAddress,
+      wallet,
+    );
+    await revokeErc20Approval(
+      getBatchConversionProxyAddress(FAUValidRequest, batchConvVersion),
+      DAITokenAddress,
```
Comment (bertux) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:182]:
```suggestion
      FAUTokenAddress,
```

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts [hunk 9]
```diff
@@ -0,0 +1,691 @@
...
+  //   AggEurUsd.sol     x  1.20
+  //   AggDaiUsd.sol     /  1.01
+  return BigNumber.from(10).pow(18).mul(amount).div(100).mul(120).div(100).mul(100).div(101);
+};
+
+describe('erc20-batch-conversion-proxy', () => {
+  beforeAll(async () => {
+    // Revoke DAI and FAU approvals
+    await revokeErc20Approval(
+      getBatchConversionProxyAddress(validEURRequest, batchConvVersion),
```
Comment (bertux) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:176]:
```suggestion
      getBatchConversionProxyAddress(DAIValidRequest, batchConvVersion),
```

**Review by bertux — CHANGES_REQUESTED** [review]

File: packages/payment-processor/src/payment/batch-proxy.ts (hunk 6 above)
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-proxy.ts:119]: Yes, I had created this function and I am the only one to use it

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts (hunk 9 above)
Comment (olivier7delf) [reply packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:176]: Indeed.
Thank the PR !

**Review by bertux — APPROVED** [review]

**Pull request merged**